File: backend/src/main/kotlin/fr/gouv/dgampa/rapportnav/infrastructure/api/bff/model/v2/Inquiry.kt
package fr.gouv.dgampa.rapportnav.infrastructure.api.bff.model.v2

import fr.gouv.dgampa.rapportnav.domain.entities.mission.v2.*
import java.time.Instant
import java.util.*

class Inquiry(
    var id: UUID? = null,
    val agentId: Int? = null,
    val serviceId: Int? = null,
    var endDateTimeUtc: Instant? = null,
    val startDateTimeUtc: Instant? = null,
    val type: String? = null,
    val origin: InquiryOriginType? = null,
    val status: InquiryStatusType? = null,
    val conclusion: InquiryConclusionType? = null,
    val missionId: Int? = null,
    val missionIdUUID: UUID? = null,
    var actions: List<MissionNavAction?> = listOf(),
    var isSignedByInspector: Boolean? = null,
    val vessel: Vessel?= null,
    var establishment: Establishment? = null
) {

    fun toInquiryEntity(): InquiryEntity {
        return InquiryEntity(
            id = id,
            type = type,
            status = status,
            origin = origin,
            agentId = agentId,
            serviceId = serviceId,
            conclusion = conclusion,
            endDateTimeUtc = endDateTimeUtc,
            startDateTimeUtc = startDateTimeUtc,
            missionId = missionId,
            missionIdUUID = missionIdUUID,
            isSignedByInspector = isSignedByInspector,
            actions = listOf(),
            vessel = vessel?.toVesselEntity(),
            establishment = establishment?.toEstablishmentEntity()

        )
    }

    companion object {
        fun fromInquiryEntity(entity: InquiryEntity?): Inquiry {
            return Inquiry(
                id = entity?.id,
                type = entity?.type,
                status = entity?.status,
                origin = entity?.origin,
                agentId = entity?.agentId,
                serviceId = entity?.serviceId,
                conclusion = entity?.conclusion,
                endDateTimeUtc = entity?.endDateTimeUtc,
                startDateTimeUtc = entity?.startDateTimeUtc,
                missionId = entity?.missionId,
                missionIdUUID = entity?.missionIdUUID,
                isSignedByInspector = entity?.isSignedByInspector,
                vessel = entity?.vessel?.let { Vessel.fromVesselEntity(it) },
                establishment = entity?.establishment?.let { Establishment.fromEstablishmentEntity(it) },
                actions = entity?.actions?.map { action -> MissionNavAction.fromMissionActionEntity(action) } ?: listOf()
            )
        }
    }
}

File: frontend/src/v2/features/common/types/inquiry.ts
import { MissionAction } from './mission-action'
import { Vessel } from './vessel-type'

export enum InquiryOriginType {
  OPPORTUNITY_CONTROL = 'OPPORTUNITY_CONTROL',
  FOLLOW_UP_CONTROL = 'FOLLOW_UP_CONTROL',
  CNSP_REPORTING = 'CNSP_REPORTING',
  OTHER_REPORTING = 'OTHER_REPORTING',
  URCEM_DEDICATED_STATION = 'URCEM_DEDICATED_STATION'
}

export enum InquiryStatusType {
  CLOSED = 'CLOSED',
  IN_PROGRESS = 'IN_PROGRESS'
}

export enum InquiryConclusionType {
  NO_FOLLOW_UP = 'NO_FOLLOW_UP',
  WITH_REPORT = 'WITH_REPORT'
}

export enum InquiryTargetType {
  VEHICLE = 'VEHICLE',
  COMPANY = 'COMPANY'
}

export interface Inquiry {
  id?: string
  type?: InquiryTargetType
  agentId?: number
  serviceId?: number
  endDateTimeUtc?: string
  startDateTimeUtc?: string
  origin?: InquiryOriginType
  status?: InquiryStatusType
  conclusion?: InquiryConclusionType
  isSignedByInspector?: boolean
  missionId?: number
  siren?: string
  missionIdUUID?: string
  actions?: MissionAction[]
  vessel?: Vessel
  isForeignEstablishment?: boolean
}
